packages/vhd-lib/src/createVhdDirectoryFromStream.js

```diff
@@ -0,0 +1,179 @@
+import { VhdDirectory } from './'
+import { BLOCK_UNUSED, FOOTER_SIZE, HEADER_SIZE, SECTOR_SIZE } from './_constants'
+import { readChunk } from '@vates/read-chunk'
+import assert from 'assert'
+import { Disposable } from 'promise-toolbox'
+import { buildFooter, buildHeader, computeBlockBitmapSize } from './Vhd/_utils'
+import { asyncEach } from '@vates/async-each'
+
+const cappedBufferConcat = (buffers, maxSize) => {
+  let buffer = Buffer.concat(buffers)
+  if (buffer.length > maxSize) {
+    buffer = buffer.slice(buffer.length - maxSize)
+  }
+  return buffer
+}
+
+async function* parse(stream) {
+  let bytesRead = 0
+
+  // handle empty space between elements
+  // ensure we read stream in order
+  async function read(offset, size) {
+    assert(bytesRead <= offset, `offset is ${offset} but we already read ${bytesRead} bytes`)
+    if (bytesRead < offset) {
+      // empty spaces
+      await read(bytesRead, offset - bytesRead)
+    }
+    const buf = await readChunk(stream, size)
+    assert.strictEqual(buf.length, size, `read ${buf.length} instead of ${size}`)
```

Comment:
Note: providing a message like you did is often better but not mandatory when using `assert.strictEqual` which shows the compared values by default :slightly_smiling_face: